Please test with Docutils 0.19b1

Hi,

We've just released Docutils 0.19b1, and intend to release 0.19 final on 05/07/2022 (a week on Tuesday). Please would you test with the pre-release, and raise any issues to me either on this issue or on the Docutils tracker?

- The release notes are at: https://docutils.sourceforge.io/RELEASE-NOTES.html#release-0-19b1-2022-06-21
- The detailed change history is at: https://docutils.sourceforge.io/HISTORY.html#release-0-19b1-2022-06-21
- You can install 0.19b1 with `python -m pip install --pre "docutils==0.19b1"` or install Docutils' master with `python -m pip install "git+https://repo.or.cz/docutils.git#subdirectory=docutils"`.

Thanks,
Adam

I ran this with docutils 0.19, and got the following test failures:

```
=================================== FAILURES ===================================
____________________ test_docutils_roles[90-pep-reference] _____________________

file_params = ParamTestData(line=90, title='pep-reference', description='(`docutils.parsers.rst.roles.pep_reference_role`):', conten...g/dev/peps/pep-0000">\n            PEP 0\n', index=8, fmt=<pytest_param_files.main.DotFormat object at 0x7ffff4fd6ad0>)
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7ffff3ed8dc0>

    @pytest.mark.param_file(FIXTURE_PATH / "docutil_roles.md")
    def test_docutils_roles(file_params, monkeypatch):
        """Test conversion of Markdown to docutils AST (before transforms are applied)."""
    
        def _apply_transforms(self):
            pass
    
        monkeypatch.setattr(Publisher, "apply_transforms", _apply_transforms)
    
        doctree = publish_doctree(
            file_params.content,
            source_path="notset",
            parser=Parser(),
        )
    
>       file_params.assert_expected(doctree.pformat(), rstrip_lines=True)
E       AssertionError: Actual does not match expected
E       --- /build/source/tests/test_renderers/fixtures/docutil_roles.md:90
E       +++ (actual)
E       @@ -1,4 +1,4 @@
E        <document source="notset">
E       
E            <paragraph>
E       
E       -        <reference refuri="http://www.python.org/dev/peps/pep-0000">
E       
E       +        <reference refuri="https://peps.python.org/pep-0000">
E       
E                    PEP 0

tests/test_renderers/test_fixtures_docutils.py:53: AssertionError
____________________ test_docutils_roles[101-rfc-reference] ____________________

file_params = ParamTestData(line=101, title='rfc-reference', description='(`docutils.parsers.rst.roles.rfc_reference_role`):', conte....org/html/rfc1.html">\n            RFC 1\n', index=9, fmt=<pytest_param_files.main.DotFormat object at 0x7ffff4fd6ad0>)
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7ffff40110c0>

    @pytest.mark.param_file(FIXTURE_PATH / "docutil_roles.md")
    def test_docutils_roles(file_params, monkeypatch):
        """Test conversion of Markdown to docutils AST (before transforms are applied)."""
    
        def _apply_transforms(self):
            pass
    
        monkeypatch.setattr(Publisher, "apply_transforms", _apply_transforms)
    
        doctree = publish_doctree(
            file_params.content,
            source_path="notset",
            parser=Parser(),
        )
    
>       file_params.assert_expected(doctree.pformat(), rstrip_lines=True)
E       AssertionError: Actual does not match expected
E       --- /build/source/tests/test_renderers/fixtures/docutil_roles.md:101
E       +++ (actual)
E       @@ -1,4 +1,4 @@
E        <document source="notset">
E       
E            <paragraph>
E       
E       -        <reference refuri="http://tools.ietf.org/html/rfc1.html">
E       
E       +        <reference refuri="https://tools.ietf.org/html/rfc1.html">
E       
E                    RFC 1


```

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -70,7 +70,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        docutils-version: ["0.16", "0.17", "0.18"]
+        docutils-version: ["0.17", "0.18", "0.19"]
 
     steps:
     - name: Checkout source

--- a/myst_parser/mdit_to_docutils/base.py
+++ b/myst_parser/mdit_to_docutils/base.py
@@ -1391,11 +1391,16 @@ def html_meta_to_nodes(
         return []
 
     try:
+        # if sphinx available
         from sphinx.addnodes import meta as meta_cls
     except ImportError:
-        from docutils.parsers.rst.directives.html import MetaBody
+        try:
+            # docutils >= 0.19
+            meta_cls = nodes.meta  # type: ignore
+        except AttributeError:
+            from docutils.parsers.rst.directives.html import MetaBody
 
-        meta_cls = MetaBody.meta  # type: ignore
+            meta_cls = MetaBody.meta  # type: ignore
 
     output = []
 

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -34,7 +34,7 @@ keywords = [
 ]
 requires-python = ">=3.7"
 dependencies = [
-    "docutils>=0.15,<0.19",
+    "docutils>=0.15,<0.20",
     "jinja2", # required for substitutions, but let sphinx choose version
     "markdown-it-py>=1.0.0,<3.0.0",
     "mdit-py-plugins~=0.3.0",

--- a/tests/test_renderers/fixtures/docutil_roles.md
+++ b/tests/test_renderers/fixtures/docutil_roles.md
@@ -93,7 +93,7 @@
 .
 <document source="notset">
     <paragraph>
-        <reference refuri="http://www.python.org/dev/peps/pep-0000">
+        <reference refuri="https://peps.python.org/pep-0000">
             PEP 0
 .
 
@@ -104,7 +104,7 @@
 .
 <document source="notset">
     <paragraph>
-        <reference refuri="http://tools.ietf.org/html/rfc1.html">
+        <reference refuri="https://tools.ietf.org/html/rfc1.html">
             RFC 1
 .
 

--- a/tests/test_renderers/test_fixtures_docutils.py
+++ b/tests/test_renderers/test_fixtures_docutils.py
@@ -50,7 +50,18 @@ def test_docutils_roles(file_params, monkeypatch):
         parser=Parser(),
     )
 
-    file_params.assert_expected(doctree.pformat(), rstrip_lines=True)
+    ptree = doctree.pformat()
+    # docutils >=0.19 changes:
+    ptree = ptree.replace(
+        'refuri="http://tools.ietf.org/html/rfc1.html"',
+        'refuri="https://tools.ietf.org/html/rfc1.html"',
+    )
+    ptree = ptree.replace(
+        'refuri="http://www.python.org/dev/peps/pep-0000"',
+        'refuri="https://peps.python.org/pep-0000"',
+    )
+
+    file_params.assert_expected(ptree, rstrip_lines=True)
 
 
 @pytest.mark.param_file(FIXTURE_PATH / "docutil_directives.md")

--- a/tests/test_sphinx/test_sphinx_builds.py
+++ b/tests/test_sphinx/test_sphinx_builds.py
@@ -299,6 +299,10 @@ def test_include_from_rst(
     )
 
 
+@pytest.mark.skipif(
+    __version_info__ < VersionInfo(0, 19, 0, "final", 0, True),
+    reason="Footnote HTML changed in docutils 0.19",
+)
 @pytest.mark.sphinx(
     buildername="html", srcdir=os.path.join(SOURCE_DIR, "footnotes"), freshenv=True
 )

--- a/tests/test_sphinx/test_sphinx_builds/test_footnotes.sphinx5.html
+++ b/tests/test_sphinx/test_sphinx_builds/test_footnotes.sphinx5.html
@@ -97,113 +97,115 @@
      </div>
     </blockquote>
     <hr class="footnotes docutils"/>
-    <aside class="footnote brackets" id="c" role="note">
-     <span class="label">
-      <span class="fn-bracket">
-       [
+    <aside class="footnote-list brackets">
+     <aside class="footnote brackets" id="c" role="note">
+      <span class="label">
+       <span class="fn-bracket">
+        [
+       </span>
+       <a href="#id1" role="doc-backlink">
+        1
+       </a>
+       <span class="fn-bracket">
+        ]
+       </span>
       </span>
-      <a href="#id1" role="doc-backlink">
-       1
-      </a>
-      <span class="fn-bracket">
-       ]
+      <p>
+       a footnote referenced first
+      </p>
+     </aside>
+     <aside class="footnote brackets" id="d" role="note">
+      <span class="label">
+       <span class="fn-bracket">
+        [
+       </span>
+       <a href="#id2" role="doc-backlink">
+        2
+       </a>
+       <span class="fn-bracket">
+        ]
+       </span>
       </span>
-     </span>
-     <p>
-      a footnote referenced first
-     </p>
-    </aside>
-    <aside class="footnote brackets" id="d" role="note">
-     <span class="label">
-      <span class="fn-bracket">
-       [
+      <p>
+       a footnote referenced in a directive
+      </p>
+     </aside>
+     <aside class="footnote brackets" id="a" role="note">
+      <span class="label">
+       <span class="fn-bracket">
+        [
+       </span>
+       <a href="#id3" role="doc-backlink">
+        3
+       </a>
+       <span class="fn-bracket">
+        ]
+       </span>
       </span>
-      <a href="#id2" role="doc-backlink">
-       2
-      </a>
-      <span class="fn-bracket">
-       ]
+      <p>
+       some footnote
+       <em>
+        text
+       </em>
+      </p>
+     </aside>
+     <aside class="footnote brackets" id="b" role="note">
+      <span class="label">
+       <span class="fn-bracket">
+        [
+       </span>
+       <a href="#id4" role="doc-backlink">
+        4
+       </a>
+       <span class="fn-bracket">
+        ]
+       </span>
       </span>
-     </span>
-     <p>
-      a footnote referenced in a directive
-     </p>
-    </aside>
-    <aside class="footnote brackets" id="a" role="note">
-     <span class="label">
-      <span class="fn-bracket">
-       [
+      <p>
+       a footnote before its reference
+      </p>
+     </aside>
+     <aside class="footnote brackets" id="id8" role="note">
+      <span class="label">
+       <span class="fn-bracket">
+        [
+       </span>
+       123
+       <span class="fn-bracket">
+        ]
+       </span>
       </span>
-      <a href="#id3" role="doc-backlink">
-       3
-      </a>
-      <span class="fn-bracket">
-       ]
+      <span class="backrefs">
+       (
+       <a href="#id5" role="doc-backlink">
+        1
+       </a>
+       ,
+       <a href="#id6" role="doc-backlink">
+        2
+       </a>
+       )
       </span>
-     </span>
-     <p>
-      some footnote
-      <em>
-       text
-      </em>
-     </p>
-    </aside>
-    <aside class="footnote brackets" id="b" role="note">
-     <span class="label">
-      <span class="fn-bracket">
-       [
+      <p>
+       multiple references footnote
+      </p>
+     </aside>
+     <aside class="footnote brackets" id="e" role="note">
+      <span class="label">
+       <span class="fn-bracket">
+        [
+       </span>
+       <a href="#id7" role="doc-backlink">
+        5
+       </a>
+       <span class="fn-bracket">
+        ]
+       </span>
       </span>
-      <a href="#id4" role="doc-backlink">
-       4
-      </a>
-      <span class="fn-bracket">
-       ]
-      </span>
-     </span>
-     <p>
-      a footnote before its reference
-     </p>
-    </aside>
-    <aside class="footnote brackets" id="id8" role="note">
-     <span class="label">
-      <span class="fn-bracket">
-       [
-      </span>
-      123
-      <span class="fn-bracket">
-       ]
-      </span>
-     </span>
-     <span class="backrefs">
-      (
-      <a href="#id5" role="doc-backlink">
-       1
-      </a>
-      ,
-      <a href="#id6" role="doc-backlink">
-       2
-      </a>
-      )
-     </span>
-     <p>
-      multiple references footnote
-     </p>
-    </aside>
-    <aside class="footnote brackets" id="e" role="note">
-     <span class="label">
-      <span class="fn-bracket">
-       [
-      </span>
-      <a href="#id7" role="doc-backlink">
-       5
-      </a>
-      <span class="fn-bracket">
-       ]
-      </span>
-     </span>
-     <p>
-      footnote definition in a block element
-     </p>
+      <p>
+       footnote definition in a block element
+      </p>
+     </aside>
     </aside>
    </section>
   </div>

